src/violation.rs

```diff
@@ -72,14 +103,42 @@ impl ViolationWithContext {
             }
         );
     }
+
+    pub fn message(&self) -> Cow<'_, str> {
+        let message_template = match &self.message_or_message_id {
+            MessageOrMessageId::Message(message) => message,
+            MessageOrMessageId::MessageId(message_id) => self
+                .rule
+                .messages
+                .as_ref()
+                .expect("No messages for rule")
+                .get(message_id)
+                .unwrap_or_else(|| panic!("Invalid message ID for rule: {message_id:?}")),
+        };
+        format_message(message_template, self.data.as_ref())
+    }
 }
 
-#[macro_export]
-macro_rules! violation {
-    ($($variant:ident => $value:expr),* $(,)?) => {
-        $crate::builder_args!(
-            $crate::ViolationBuilder,
-            $($variant => $value),*,
-        )
+fn format_message<'a>(message_template: &'a str, data: Option<&ViolationData>) -> Cow<'a, str> {
```

Comment:
I didn't know what to reach for for this (the `{{ ... }}` "interpolation") so I hand-rolled it

Slightly proud that I apparently did it correctly the first try